// gome-bridge/webpack.config.js
/*
 * @Author: zhaoye 
 * @Date: 2017-01-03 19:46:36
 * @Last Modified by: zhaoye
 * @Last Modified time: 2017-09-05 23:39:42
 */
var path = require('path');
var webpack = require('webpack');

function conf(target){
    var filename = ''
    return{
        entry: './index.js',
        output: {
            path: path.resolve(__dirname, 'dist'),
            filename: (target == 'gome-bridge') ? 'gome-bridge.js' : (target == 'var' ? 'gomeBridge.js' : 'gomeBridge.umd.js') ,
            libraryTarget: target == 'gome-bridge' ? 'umd' : target,
            library: ["$g"]
        },
        externals: {
            'gome-bridge-core': 'GomeJSBridge'  
        },
        vue: {
            loaders: {
                js: 'babel',
            }
        },
        imageWebpackLoader: {
            mozjpeg: {
                quality: 70
            },
            pngquant:{
                quality: "70",
                speed: 1
            },
        },
        module: {
            loaders: [
                {
                    test: /\.vue$/,
                    loader: 'vue',
                },
                {test: /\.less$/, loader: 'style!css!less'},
                {test: /\.html$/, loader: 'html'},
                { test: /\.hbs$/, loader: "handlebars" },
                {test: /\.css$/, loader: 'style!css!less'},
                {
                    test: /\.js$/,
                    loader: 'babel-loader'
                }
            ]
        },
        plugins: [
            new webpack.optimize.UglifyJsPlugin({
                compress: {
                    sequences     : false,  // join consecutive statemets with the “comma operator”
                    properties    : false,  // optimize property access: a["foo"] → a.foo
                    dead_code     : false,  // discard unreachable code
                    drop_debugger : false,  // discard “debugger” statements
                    unsafe        : false, // some unsafe optimizations (see below)
                    conditionals  : false,  // optimize if-s and conditional expressions
                    comparisons   : false,  // optimize comparisons
                    evaluate      : false,  // evaluate constant expressions
                    booleans      : false,  // optimize boolean expressions
                    loops         : false,  // optimize loops
                    unused        : false,  // drop unused variables/functions
                    hoist_funs    : false,  // hoist function declarations
                    hoist_vars    : false, // hoist variable declarations
                    if_return     : false,  // optimize if-s followed by return/continue
                    join_vars     : false,  // join var declarations
                    cascade       : false,  // try to cascade `right` into `left` in sequences
                    side_effects  : false,  // drop side-effect-free statements
                    warnings      : false,  // warn about potentially dangerous optimizations/code
                    global_defs   : {}     // global definitions
                },
                comments: false,
                exclude: /x\.js/,
                mangle: {
                    except: ['gome-utils-base64']
                }
            }),
            new webpack.SourceMapDevToolPlugin({
                test: [/\.js/,/\.vue/],
                filename: '[file].map',
                append: '//@ sourceMappingURL=[url]',
                moduleFilenameTemplate: "debug:///[resource-path]",
                fallbackModuleFilenameTemplate: "debug:///[resource-path]",
            }),
        ]
    }
} 




module.exports = [conf('var'), conf('umd'), conf('gome-bridge')]
