// gome-bridge/webpack.config.dev.js
/*
 * @Author: zhaoye 
 * @Date: 2017-01-03 19:46:36
 * @Last Modified by: zhaoye
 * @Last Modified time: 2017-02-22 14:12:22
 */
var path = require('path');
var webpack = require('webpack');

module.exports = {
    entry: './test/index.js',
    output: {
        path: path.resolve(__dirname, 'test'),
        filename: 'test.js',
    },
    vue: {
        loaders: {
            js: 'babel',
        }
    },
    imageWebpackLoader: {
        mozjpeg: {
            quality: 70
        },
        pngquant:{
            quality: "70",
            speed: 1
        },
    },
    module: {
        loaders: [
            {
                test: /\.vue$/,
                loader: 'vue',
            },
            {test: /\.less$/, loader: 'style!css!less'},
            {test: /\.html$/, loader: 'html'},
            { test: /\.hbs$/, loader: "handlebars" },
            {test: /\.css$/, loader: 'style!css!less'},
            {
                test: /\.js$/,
                loader: 'babel-loader'
            }
        ]
    },
    plugins: [
        new webpack.optimize.DedupePlugin(),
        new webpack.optimize.UglifyJsPlugin({
            compress: {
                warnings: false
            },
            comments: false
        }),
        new webpack.SourceMapDevToolPlugin({
            test: [/\.js/,/\.vue/],
            filename: '[file].map',
            append: '//@ sourceMappingURL=[url]',
            moduleFilenameTemplate: "debug:///[resource-path]",
            fallbackModuleFilenameTemplate: "debug:///[resource-path]",
        }),
    ]
    
}
