// gome-utils-env/index.js
/*
 * @Author: zhaoye 
 * @Date: 2017-01-03 17:17:04 
 * @Last Modified by: liuxiaocong
 * @Last Modified time: 2018-03-01 17:40:55
 */
import {location, navigator} from 'gome-utils-bom'
/**
 * @param Object 
 * 通过ua判断的环境变量
 */
let env = {
    /****按终端分*****/
    //wap，浏览器中
    __wap      :  false,
    //app中，包含网页嵌套和混合app本地页面
    __app      :  '',
    //混合app
    __hybrid   :  false,
    //微信
    __wechat   :  false,
    // 微博
    __weibo    :  false,
    // qq
    __qq       :  false,
    // pc端
    __pc       :  false,
    // 移动端
    __mobile   :  false,
    /****特殊容器*****/
    //国美在线容器
    __gome     :  false,
    //美信容器
    __gomeplus :  false,
    __plusWebview:  false,
    // 国美管家容器
    __gomegj :  false,
    // 美店APP容器
    __gomemeidian : false,
    // 国美备份容器
    __gomebackup : false,
    /****按环境分*****/
    //开发模式（混合app专用）
    __dev      :  false,
    //线上环境
    __live     :  false,
    //uat环境
    __uat      :  false,
    //预生产
    __tslive   :  false,
    //什么都判断不出来    
    __unknown  :  false,
    backup_app    : null,
    backup_wap    : null,
    backup_live   : null,
    backup_uat    : null,
    backup_tslive : null,
}
Object.defineProperties(env, {
    /****按终端分*****/
    //wap，浏览器中
    wap: {
        get () {
            if(this.backup_wap != null){
                return this.backup_wap
            }
            // 只要不是混合app，都是
            if (location.getProtocol().match(/^file/)) {
                this.__wap = false
            } else if (location.getProtocol().match(/^http/)){
                this.__wap = true
            }
            return this.__wap
        },
        set (value) {
            if(value)
                this.backup_wap = value
        },
    },
    //app中，包含网页嵌套和混合app本地页面
    app: {
        get () {
            if(this.backup_app != null){
                return this.backup_app
            }
            if (location.getHref().match(/^file\:\/\//) || this.dev) {
                this.__app = true
            } else if (navigator.getUserAgent().toLowerCase().match(/gome\/|gomeplus\/|gomegj\/|gomemeidian\/|gomebackup\//)) {
                this.__app = true
            } else {
                this.__app = false
            }
            return this.__app
        },
        set (value) {
            if(value) {
                this.backup_app = value
            }
        },
    },
    //混合app
    hybrid: {
        get () {
            // 判断是否开发模式（仅混合app）
            if (location.getHref().match(/^file\:\/\//) || this.dev) {
                this.__hybrid = true
            } else {
                this.__hybrid = false
            }
            return this.__hybrid
        },
    },
    //微信
    wechat: {
        get () {
            if (navigator.getUserAgent().match(/MicroMessenger/i)) {
                this.__wechat = true
            } else {
                this.__wechat = false
            }
            return this.__wechat
        },
    },
    weibo: {
        get () {
            if (navigator.getUserAgent().match(/Weibo/i)) {
                this.__weibo = true
            } else {
                this.__weibo = false
            }
            return this.__weibo
        },
    },
    qq: {
        get () {
            if (navigator.getUserAgent().match(/QQ/i)) {
                this.__qq = true
            } else {
                this.__qq = false
            }
            return this.__qq
        },
    },
    mobile: {
        get () {
            if (navigator.getUserAgent().match(/(iPhone|iPod|Android|ios|iPad)/i)) {
                this.__mobile = true
            } else {
                this.__mobile = false
            }
            return this.__mobile
        },
    },
    pc: {
        get () {
            if (!navigator.getUserAgent().match(/(iPhone|iPod|Android|ios|iPad)/i)) {
                this.__pc = true
            } else {
                this.__pc = false
            }
            return this.__pc
        },
    },
    /****特殊容器*****/
    //国美在线容器
    gome: {
        get () {
            if (this.app && location.getHost().match(/gome\.com\.cn|atguat\.com\.cn|tslive\.com\.cn/)) {
                this.__gome = true
            } else {
                this.__gome = false
            }
            return this.__gome
        },
    },
    //美信容器
    gomeplus: {
        get () {
            if (this.app && location.getHost().match(/gomeplus\.com|uatplus\.com|tsliveplus\.com\.cn/)) {
                this.__gomeplus = true
            } else {
                this.__gomeplus = false
            }
            return this.__gomeplus
        },
    },
    gomebackup:{
        get () {
            if (navigator.getUserAgent().match(/gomebackup\//)) {
                this.__gomebackup = true
            } else {
                this.__gomebackup = false
            }
            return this.__gomebackup
        },
    },
    plusWebview: {

    },
    //国美管家
    gomegj : {
        get () {
            if(navigator.getUserAgent().match(/gomegj/i)){
                this.__gomegj = true
            } else {
                this.__gomegj = false
            }
            return this.__gomegj
        },
    },
     //国美美店
     gomemeidian : {
        get () {
            if(navigator.getUserAgent().toLowerCase().match(/gomemeidian/i)){
                this.__gomemeidian = true
            } else {
                this.__gomemeidian = false
            }
            return this.__gomemeidian
        },
    },
    /****按环境分*****/
    // 开发模式（混合app专用）
    dev: {
        get () {
            // 判断是否开发模式（仅混合app）
            if (navigator.getUserAgent().match(/dev\b/)) {
                this.__dev = true
            } else {
                this.__dev = false
            }
            return this.__dev
        },
    },
    //线上环境
    live: {
        get () {
            if(this.backup_live != null){
                return this.backup_live
            }
            if (location.getHost().match(/gome\.com\.cn|gomeplus\.com/)) {
                this.__live = true
            } else {
                this.__live = false
            }
            return this.__live
        },
        set (value) {
            if(value) {
                this.backup_live = value
            }
        },
    },
    //uat环境
    uat: {
        get () {
            if(this.backup_uat != null){
                return this.backup_uat
            }
            if (location.getHost().match(/uat.*\.com/)) {
                this.__uat = true
            } else {
                this.__uat = false
            }
            return this.__uat
        },
        set (value) {
            if(value) {
                this.backup_uat = value
            }
        },
    },
    //预生产
    tslive: {
        get () {
            if(this.backup_tslive != null){
                return this.backup_tslive
            }
            if (location.getHost().match(/tslive(plus)?\.com(\.cn)?/)) {
                this.__tslive = true
            } else {
                this.__tslive = false
            }
            return this.__tslive
        },
        set (value) {
            if(value) {
                this.backup_tslive = value
            }
        },
    },
    //什么都判断不出来    
    unknown: {
        get () {
            if (!(location.getHost().match(/gome\.com\.cn|gomeplus\.com/)) &&
                !(location.getHost().match(/uat.*\.com/)) && 
                !(location.getHost().match(/tslive(plus)?\.com(\.cn)?/))) {
                this.__unknown = true
            } else {
                this.__unknown = false
            }
            return this.__unknown
        },
    },
})

//系统判断
let system = {
    __android : false,
    __ios     : false
}
Object.defineProperties(system, {
    android: {
        get () {
            if (navigator.getUserAgent().match(/android/i)) {
                this.__android = true
            } else {
                this.__android = false
            }
            return this.__android
        },
    },
    ios: {
        get () {
            if (navigator.getUserAgent().match(/iPhone|iPad/i)) {
                this.__ios = true
            } else {
                this.__ios = false
            }
            return this.__ios
        },
    },
})

if(env.unknown && window.PackConfig){
    env[PackConfig.PLATFORM.toLowerCase()] = true
    env[PackConfig.ENV.toLowerCase()] = true
}

//第三方环境复写
if(window.GomeBridgePreConfig && window.GomeBridgePreConfig.env){
    for(var key in GomeBridgePreConfig.env){
        env[key] = GomeBridgePreConfig.env[key]
    }
}

export {
    env,
    system
}

export default {
    env,
    system
}
