// gome-bridge-core/webpack.config.cordova.js
/*
 * @Author: liuxiaocong 
 * @Date: 2017-08-22 16:40:35 
 * @Last Modified by: liuxiaocong
 * @Last Modified time: 2017-08-22 19:20:34
 */
var path = require('path')
var webpack = require('webpack')
var UglifyJsPlugin = require('uglifyjs-webpack-plugin');

function build(uglify){
    if(uglify){
        var config = {
            // entry:['./src/cordova.js','./src/init.js','./src/bridge.port.js'],
            entry:'./example/cordova.unit.less.js',
            output: {
                path: path.resolve(__dirname, 'exampleDist'),
                filename: 'GomeJSBridge.min.js',
                libraryTarget: 'umd',
                library:'GomeJSBridge'
            },
            module: {
                loaders: [
                    {
                        test: /\.js$/,
                        loader: 'babel-loader'
                    }
                ]
            },
            devtool: 'source-map',
            plugins: [
                new UglifyJsPlugin({
                    mangle: {
                        // 跳过这些
                        // except : ['require', 'exports', 'module','GomeJSBridge','_cordovaNative']
                    },
                })
            ]
        }
        return config
    }else{
        var config = {
            // entry:['./src/cordova.js','./src/init.js','./src/bridge.port.js'],
            entry:'./src/index.js',
            output: {
                path: path.resolve(__dirname, 'dist'),
                filename: 'GomeJSBridge.js',
                libraryTarget: 'umd',
                library:'GomeJSBridge'
            },
            module: {
                loaders: [
                    {
                        test: /\.js$/,
                        loader: 'babel-loader'
                    }
                ]
            },
            devtool: 'source-map'
        }
        return config
    }
}
module.exports = [build(true)]
