// gome-bridge-core/webpack.config.js
/*
 * @Author: liuxiaocong 
 * @Date: 2017-07-19 20:46:24 
 * @Last Modified by: liuxiaocong
 * @Last Modified time: 2017-08-23 10:48:43
 */
var path = require('path')
var webpack = require('webpack')
var UglifyJsPlugin = require('uglifyjs-webpack-plugin');

function build(uglify){
    if(uglify){
        var config = {
            // entry:['./src/cordova.js','./src/init.js','./src/bridge.port.js'],
            entry:'./example/cordova.unit.less.js',
            output: {
                path: path.resolve(__dirname, 'dist'),
                filename: 'GomeJSBridge.min.js',
                libraryTarget: 'umd',
                library:'GomeJSBridge'
            },
            module: {
                loaders: [
                    {
                        test: /\.js$/,
                        loader: 'babel-loader'
                    }
                ]
            },
            devtool: 'source-map',
            plugins: [
                new UglifyJsPlugin({

                    mangle: {
                        // 跳过这些
                        // except : ['require', 'exports', 'module']
                    },
                    // beautify : false,
                    // sourceMap : true
                })
            ]
        }
        return config
    }else{
        var config = {
            // entry:['./src/cordova.js','./src/init.js','./src/bridge.port.js'],
            entry:'./example/cordova.unit.less.js',
            output: {
                path: path.resolve(__dirname, 'dist'),
                filename: 'GomeJSBridge.js',
                libraryTarget: 'umd',
                library:'GomeJSBridge'
            },
            module: {
                loaders: [
                    {
                        test: /\.js$/,
                        loader: 'babel-loader'
                    }
                ]
            },
            devtool: 'source-map'
        }
        return config
    }
}
module.exports = [build(true),build()]
